// @lib/autodts-generator/index.js
/// <reference path="./typings/tsd" />
var fs = require('fs');
var mkdirp = require('mkdirp');
var os = require('os');
var pathUtil = require('path');
var Promise = require('bluebird');
var ts = require('typescript');
var filenameToMid = (function () {
    if (pathUtil.sep === '/') {
        return function (filename) {
            return filename;
        };
    }
    else {
        var separatorExpression = new RegExp(pathUtil.sep.replace('\\', '\\\\'), 'g');
        return function (filename) {
            return filename.replace(separatorExpression, '/');
        };
    }
})();
function getError(diagnostics) {
    var message = 'Declaration generation failed';
    diagnostics.forEach(function (diagnostic) {
        var position = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start);
        message +=
            ("\n" + diagnostic.file.fileName + "(" + (position.line + 1) + "," + (position.character + 1) + "): ") +
                ("error TS" + diagnostic.code + ": " + diagnostic.messageText);
    });
    var error = new Error(message);
    error.name = 'EmitterError';
    return error;
}
function getFilenames(baseDir, files) {
    return files.map(function (filename) {
        var resolvedFilename = pathUtil.resolve(filename);
        if (resolvedFilename.indexOf(baseDir) === 0) {
            return resolvedFilename;
        }
        return pathUtil.resolve(baseDir, filename);
    });
}
function processTree(sourceFile, replacer) {
    var code = '';
    var cursorPosition = 0;
    function skip(node) {
        cursorPosition = node.end;
    }
    function readThrough(node) {
        code += sourceFile.text.slice(cursorPosition, node.pos);
        cursorPosition = node.pos;
    }
    function visit(node, parent) {
        readThrough(node);
        var replacement = replacer(node, parent);
        if (replacement != null) {
            code += replacement;
            skip(node);
        }
        else {
            parent = node;
            ts.forEachChild(node, function (node) { visit(node, parent); });
        }
    }
    visit(sourceFile, null);
    code += sourceFile.text.slice(cursorPosition);
    return code;
}
function generate(options, sendMessage) {
    if (sendMessage === void 0) { sendMessage = function () { }; }
    var baseDir = pathUtil.resolve(options.baseDir);
    var eol = options.eol || os.EOL;
    var nonEmptyLineStart = new RegExp(eol + '(?!' + eol + '|$)', 'g');
    var indent = options.indent === undefined ? '\t' : options.indent;
    var target = options.target || 2 /* Latest */;
    var compilerOptions = {
        declaration: true,
        module: 1 /* CommonJS */,
        target: target
    };
    var filenames = getFilenames(baseDir, options.files);
    var excludesMap = {};
    options.excludes && options.excludes.forEach(function (filename) {
        excludesMap[filenameToMid(pathUtil.resolve(baseDir, filename))] = true;
    });
    mkdirp.sync(pathUtil.dirname(options.out));
    var output = fs.createWriteStream(options.out, { mode: parseInt('644', 8) });
    var outputRef;
    var outputString = '';
    var outputRefPath;
    var outputRefString = '';
    var wroteToOutputRef = false;
    if (options.out.slice(-5) === '.d.ts') {
        outputRefPath = options.out.slice(0, -5) + '.ref.d.ts';
        outputRef = fs.createWriteStream(outputRefPath, { mode: parseInt('644', 8) });
    }
    var host = ts.createCompilerHost(compilerOptions);
    var program = ts.createProgram(filenames, compilerOptions, host);
    function writeFile(filename, data, writeByteOrderMark) {
        // Compiler is emitting the non-declaration file, which we do not care about
        if (filename.slice(-5) !== '.d.ts') {
            return;
        }
        writeDeclaration(ts.createSourceFile(filename, data, target, true));
    }
    return new Promise(function (resolve, reject) {
        var closeResolver = function () { resolve(undefined); };
        output.on('close', closeResolver);
        output.on('error', reject);
        if (options.externs) {
            options.externs.forEach(function (path) {
                sendMessage("Writing external dependency " + path);
                outputString += ("/// <reference path=\"" + path + "\" />") + eol;
            });
        }
        program.getSourceFiles().some(function (sourceFile) {
            // Source file is a default library, or other dependency from another project, that should not be included in
            // our bundled output
            if (pathUtil.normalize(sourceFile.fileName).indexOf(baseDir) !== 0) {
                return;
            }
            if (excludesMap[filenameToMid(pathUtil.normalize(sourceFile.fileName))]) {
                return;
            }
            sendMessage("Processing " + sourceFile.fileName);
            // Source file is already a declaration file so should does not need to be pre-processed by the emitter
            if (sourceFile.fileName.slice(-5) === '.d.ts') {
                writeDeclaration(sourceFile);
                return;
            }
            var emitOutput = program.emit(sourceFile, writeFile);
            if (emitOutput.emitSkipped || emitOutput.diagnostics.length > 0) {
                reject(getError(emitOutput.diagnostics
                    .concat(program.getSemanticDiagnostics(sourceFile))
                    .concat(program.getSyntacticDiagnostics(sourceFile))
                    .concat(program.getDeclarationDiagnostics(sourceFile))));
                return true;
            }
        });
        if (options.main) {
            outputString += ("declare module '" + options.name + "' {") + eol + indent;
            outputString += ("import main = require('" + options.main + "');") + eol + indent;
            outputString += 'export = main;' + eol;
            outputString += '}' + eol;
            sendMessage("Aliased main module " + options.name + " to " + options.main);
        }
        if (wroteToOutputRef) {
            outputString = (("/// <reference path=\"" + outputRefPath + "\" />") + eol) + outputString;
        }
        output.write(outputString);
        // output.end() will complete our promise, if we need to delete an empty outputRef file then don't complete
        // until unlink has completed, otherwise complete now.
        if (outputRef) {
            if (wroteToOutputRef) {
                outputRef.write(outputRefString);
            }
            outputRef.end();
            if (!wroteToOutputRef) {
                fs.unlink(outputRefPath, function (err) {
                    if (err) {
                        sendMessage("Error on unlink of '\"" + outputRefPath + "\"': \"" + err + "\"");
                        // Don't want our close listener to call resolve, as we are about to call reject
                        output.removeListener('close', closeResolver);
                        output.end();
                        reject(err);
                    }
                    else {
                        output.end();
                    }
                });
            }
            else {
                output.end();
            }
        }
        else {
            output.end();
        }
    });
    function writeDeclaration(declarationFile) {
        var filename = declarationFile.fileName;
        var sourceModuleId = options.name + filenameToMid(filename.slice(baseDir.length, -5));
        if (declarationFile.externalModuleIndicator) {
            if (outputRefPath) {
                declarationFile.referencedFiles.forEach(function (refPath) {
                    var refFullPath = pathUtil.resolve(pathUtil.dirname(pathUtil.resolve(baseDir, filename)), refPath.fileName);
                    var refRelPath = filenameToMid(pathUtil.relative(pathUtil.dirname(outputRefPath), refFullPath));
                    outputRefString += ("/// <reference path=\"" + refRelPath + "\" />") + eol;
                    wroteToOutputRef = true;
                });
            }
            outputString += 'declare module \'' + sourceModuleId + '\' {' + eol + indent;
            var content = processTree(declarationFile, function (node, parent) {
                if (node.kind === 230 /* ExternalModuleReference */) {
                    var expression = node.expression;
                    if (expression.text.charAt(0) === '.') {
                        return ' require(\'' + filenameToMid(pathUtil.join(pathUtil.dirname(sourceModuleId), expression.text)) + '\')';
                    }
                }
                else if (node.kind === 120 /* DeclareKeyword */) {
                    return '';
                }
                else if (node.kind === 9 /* StringLiteral */ &&
                    (parent.kind === 226 /* ExportDeclaration */ || parent.kind === 220 /* ImportDeclaration */)) {
                    var text = node.text;
                    if (text.charAt(0) === '.') {
                        return " '" + filenameToMid(pathUtil.join(pathUtil.dirname(sourceModuleId), text)) + "'";
                    }
                }
            });
            outputString += content.replace(nonEmptyLineStart, '$&' + indent);
            outputString += eol + '}' + eol;
        }
        else {
            outputString += declarationFile.text;
        }
    }
}
exports.generate = generate;
